pom.xml

```diff
@@ -357,81 +356,20 @@
           </excludes>
         </configuration>
       </plugin>
-      <plugin>
-        <groupId>org.apache.maven.plugins</groupId>
-        <artifactId>maven-shade-plugin</artifactId>
-        <version>3.2.4</version>
-        <executions>
-          <!-- STEP 2: execute maven shade to rename bytebuddy -->
-          <execution>
-            <id>bytebuddy</id>
-            <phase>package</phase>
-            <goals>
-              <goal>shade</goal>
-            </goals>
-            <configuration>
-              <createDependencyReducedPom>true</createDependencyReducedPom>
-              <relocations>
-                <relocation>
-                  <pattern>net.bytebuddy</pattern>
-                  <shadedPattern>org.assertj.core.internal.bytebuddy</shadedPattern>
-                  <excludes>
-                    <exclude>net/bytebuddy/experimental</exclude>
-                  </excludes>
-                </relocation>
-              </relocations>
-              <filters>
-                <filter>
-                  <artifact>net.bytebuddy:*</artifact>
-                  <excludes>
-                    <exclude>META-INF/**</exclude>
-                  </excludes>
-                </filter>
-              </filters>
-            </configuration>
-          </execution>
-        </executions>
-      </plugin>
-      <!-- Unpack the shaded jar so that the maven-bundle-plugin can analyze the packages to import and to export -->
-      <plugin>
-        <groupId>org.apache.maven.plugins</groupId>
-        <artifactId>maven-dependency-plugin</artifactId>
-        <executions>
-          <!-- STEP 3: unpack the shaded classes back into the build dir -->
-          <execution>
-            <id>unpack-shade</id>
-            <phase>package</phase>
-            <goals>
-              <goal>unpack</goal>
-            </goals>
-            <configuration>
-              <artifactItems>
-                <artifactItem>
-                  <groupId>${project.groupId}</groupId>
-                  <artifactId>${project.artifactId}</artifactId>
-                  <version>${project.version}</version>
-                </artifactItem>
-              </artifactItems>
-              <outputDirectory>${project.build.directory}/classes</outputDirectory>
-            </configuration>
-          </execution>
-        </executions>
-      </plugin>
       <plugin>
         <groupId>biz.aQute.bnd</groupId>
         <artifactId>bnd-maven-plugin</artifactId>
         <version>${bnd.version}</version>
         <executions>
-          <!-- STEP 4: analyze & collect OSGi metadata -->
           <execution>
-            <id>bundle-after-unpack-shaded</id>
-            <phase>package</phase>
+            <id>bnd-process</id>
             <goals>
               <goal>bnd-process</goal>
             </goals>
             <configuration>
               <bnd><![CDATA[
                 Import-Package:\
+                  net.bytebuddy.*,\
```

Comment:
You *could* leave this line out - then the import of `net.bytebuddy.*` would be covered by the default catch-all below and become optional. As I understand it, BB is only needed for a subset of AssertJ functionality (assumptions, soft assertions, and...?), so one could make a case for possibly making it optional as then those who aren't using those features wouldn't need to install BB.

In that case, something that would need to be considered is how to handle `ClassNotFoundException`/`NoClassDefFoundError` at runtime in a user-friendly manner if BB is not installed. Actually, this is handling that we might want to consider having in-place for non-OSGi usage too in case BB is not on the classpath.

Reply:
If I understand it correctly, the current config would force OSGi users to install Byte Buddy even when BB related features are not used. Is the installation a manual step or does it happen automatically? My OSGi experience is close to zero so please forgive the dumb question 🙂 

Thinking about the non-OSGi world, I would keep BB as a transitive dependency to make this change as transparent as possible. Otherwise, users would need to know that from now on soft assertions and assumptions require a manual declaration of BB to work properly, which is not really nice.

Reply:
> If I understand it correctly, the current config would force OSGi users to install Byte Buddy even when BB related features are not used.

Correct. 

> Is the installation a manual step or does it happen automatically?

If it has been declared as a non- optional dependency, then it is added automatically during the OSGi resolve process (at least, it is automatic in any sanely configured OSGi build). The AssertJ bundle will fail to run if BB is not present. 

If it is an optional dependency, then you need to add it explicitly. If you don't, then the AssertJ bundle will still run - but you'll get a ClassNotFoundException or a NoClassDefFoundError at runtime if you try and use any part of the code that requires BB.

> My OSGi experience is close to zero so please forgive the dumb question 🙂
> 

No such thing as a dumb question. 😀

> Thinking about the non-OSGi world, I would keep BB as a transitive dependency to make this change as transparent as possible. Otherwise, users would need to know that from now on soft assertions and assumptions require a manual declaration of BB to work properly, which is not really nice.

If that is the plan then best to keep the OSGi dependency policy the same and leave the dependency as non-optional.

The ideal solution is to factor out separate module(s) for soft assertions and assumptions. Then the core module can be declared without the BB dependencies, and the new module(s) with. Users can pick the modules they want.

Reply:
I'd say we keep the dependency as non-optional for the current change and we might evaluate this topic again with AssertJ 4.